pkg/pool/pool_test.go

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pool
+
+import (
+	"errors"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+)
+
+func TestParallelismNoErrors(t *testing.T) {
+	tests := []struct {
+		name string
+		size int
+		work int
+	}{{
+		name: "single threaded",
+		size: 1,
+		work: 3,
+	}, {
+		name: "three workers",
+		size: 3,
+		work: 10,
+	}, {
+		name: "ten workers",
+		size: 10,
+		work: 100,
+	}}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			// m guards max
+			var m sync.Mutex
+			var max int32
+			var active int32
+
+			// Use our own waitgroup to ensure that the work
+			// can all complete before we block on the error
+			// result.
+			wg := &sync.WaitGroup{}
+
+			worker := func() error {
+				defer wg.Done()
+				na := atomic.AddInt32(&active, 1)
+				defer atomic.AddInt32(&active, -1)
+
+				func() {
+					m.Lock()
+					defer m.Unlock()
+					if max < na {
+						max = na
+					}
+				}()
+
+				time.Sleep(10 * time.Millisecond)
```

Comment:
Please comment why do we need the delay.

Reply:
Done.